reactive/docker.py

```diff
@@ -257,55 +254,88 @@ def install_from_upstream_apt():
-        check_call(split(cmd))
+        add_apt_key(key)
 
     # Get the package architecture (amd64), not the machine hardware (x86_64)
-    arch = check_output(split('dpkg --print-architecture'))
-    arch = arch.decode('utf-8').rstrip()
+    architecture = arch()
     # Get the lsb information as a dictionary.
     lsb = host.lsb_release()
     code = lsb['DISTRIB_CODENAME']
     rel = lsb['DISTRIB_RELEASE']
+    ubuntu = str(lsb['DISTRIB_ID']).lower()
     dockurl = "https://download.docker.com/linux/ubuntu"
     nvidurl = 'https://nvidia.github.io'
     repo = 'stable'
 
     deb = list()
-    deb.append('deb [arch={0}] {1} {2} {3}'.format(arch, dockurl, code, repo))
+    deb.append('deb [arch={0}] {1} {2} {3}'.format(architecture,
+                                                   dockurl, code, repo))
     for i in ['libnvidia-container',
               'nvidia-container-runtime',
               'nvidia-docker']:
-        deb.append('deb {0}/{1}/ubuntu{2}/{3} /'.format(nvidurl, i, rel, arch))
+        deb.append('deb {0}/{1}/ubuntu{2}/{3} /'.format(nvidurl,
+                                                        i, rel, architecture))
+
+    write_docker_sources(deb)
+
+    install_cuda_drivers_repo(architecture, rel, ubuntu)
+
+    apt_update(fatal=True)
+    # actually install the required packages docker-ce nvidia-docker2
+
+    docker_ce = hookenv.config('docker-ce-package')
+    nvidia_docker2 = hookenv.config('nvidia-docker-package')
+    nv_container_runtime = hookenv.config('nvidia-container-runtime-package')
+    apt_install(['cuda-drivers', docker_ce, nvidia_docker2,
+                 nv_container_runtime], fatal=True)
+
+
+def install_cuda_drivers_repo(architecture, rel, ubuntu):
+    ''' Install cuda drivers this is xenial only.
+     We want to install cuda-drivers only this means that the
+     cuda version plays no role. Any repo will do. '''
+    key_file = '7fa2af80.pub'
+    # distribution should be something like ubuntu1604
+    distribution = '{}{}'.format(ubuntu, str(rel).replace('.', ''))
+    repo_path = 'developer.download.nvidia.com/compute/' \
+                'cuda/repos/{}/x86_64'.format(distribution)
+    cmd = 'apt-key adv --fetch-keys http://{}/{}'.format(repo_path, key_file)
+    check_call(split(cmd))
+    cuda_repo_version = config('cuda_repo')
+    cuda_repo_pkg = 'cuda-repo-{}_{}_{}.deb'.format(distribution,
+                                                    cuda_repo_version,
+                                                    architecture)
+    repo_url = 'https://{}/{}'.format(repo_path, cuda_repo_pkg)
+    r = requests.get(repo_url)
+    r.raise_for_status()
+    with open(cuda_repo_pkg, "wb") as repo_file:
+        for chunk in r.iter_content(chunk_size=1024):
+            repo_file.write(chunk)
+    r.close()
+    cmd = 'dpkg -i  {}'.format(cuda_repo_pkg)
+    check_call(split(cmd))
+
 
+def write_docker_sources(deb):
```

Comment:
Any reason to have this new method versus using `add_source`?

https://github.com/juju/charm-helpers/blob/master/charmhelpers/fetch/ubuntu.py#L316

If you use the charm helper, you could remove this method and replace calls with:

```
fetch.add_source(source=deb, key='XXXXXXXXX', fail_invalid=True)
```

Two caveats:
- i *think* the charmhelper updates `/etc/apt/sources.list`, so you would not have a separate `docker.list` in `sources.list.d`.
- i'm not sure that you current `deb` format will work; the `[arch]` piece in `deb [arch] url` doesn't look like it will [match](https://github.com/juju/charm-helpers/blob/master/charmhelpers/fetch/ubuntu.py#L377), so you may have to drop it (or update the charmhelper regex).

If the above caveats are why you chose to handle apt repos yourself, very well.  I wanted to throw out the `add_source` options just in case.

Reply:
The reason for this approach is that this was already there (I think from the time Matt was with us and then the recent patch from Chris). I could give it a try but it will take me more time to test.